linkerd/app/outbound/src/metrics/transport.rs

```diff
@@ -0,0 +1,262 @@
+        match self {
+            Self::Forbidden => write!(f, "forbidden"),
+            Self::InvalidBackend => write!(f, "invalid_backend"),
+            Self::InvalidPolicy => write!(f, "invalid_policy"),
+            Self::Unexpected => write!(f, "unexpected"),
+        }
+    }
+}
+
+// === impl ConnectionsClosedLabels ===
+
+impl<L> EncodeLabelSetMut for ConnectionsClosedLabels<L>
+where
+    L: Clone + Hash + Eq + EncodeLabelSetMut + Debug + Send + Sync + 'static,
+{
+    fn encode_label_set(&self, enc: &mut LabelSetEncoder<'_>) -> std::fmt::Result {
+        self.labels.encode_label_set(enc)?;
+        match self.error {
+            Some(error) => ("error", error.to_string()).encode(enc.encode_label())?,
+            None => ("error", "").encode(enc.encode_label())?,
+        }
+
+        Ok(())
+    }
+}
+
+impl<L> EncodeLabelSet for ConnectionsClosedLabels<L>
+where
+    L: Clone + Hash + Eq + EncodeLabelSetMut + Debug + Send + Sync + 'static,
+{
+    fn encode(&self, mut enc: LabelSetEncoder<'_>) -> std::fmt::Result {
+        self.encode_label_set(&mut enc)
+    }
+}
+
+// === impl NewTransportRouteMetrics ===
+
+impl<N, L: Clone> NewTransportRouteMetrics<N, L> {
+    pub fn layer(
+        family: TransportRouteMetricsFamily<L>,
+    ) -> impl layer::Layer<N, Service = Self> + Clone {
+        layer::mk(move |inner| Self {
+            inner,
+            family: family.clone(),
+        })
+    }
+}
+
+impl<T, N, L, S> NewService<T> for NewTransportRouteMetrics<N, L>
+where
+    N: NewService<T, Service = S>,
+    L: Clone + Hash + Eq + EncodeLabelSetMut + Debug + Send + Sync + 'static,
+    T: Param<L> + Clone,
+{
+    type Service = TransportRouteMetricsService<S>;
+
+    fn new_service(&self, target: T) -> Self::Service {
+        let labels: L = target.param();
+        let metrics = self.family.metrics(labels);
+        let svc = self.inner.new_service(target);
+        TransportRouteMetricsService::new(svc, metrics)
+    }
+}
+
+// === impl TransportRouteMetricsService ===
+
+impl<S> TransportRouteMetricsService<S> {
+    fn new(inner: S, metrics: TransportRouteMetrics) -> Self {
+        Self { inner, metrics }
+    }
+}
+
+impl<I, S> Service<I> for TransportRouteMetricsService<S>
+where
+    I: io::AsyncRead + io::AsyncWrite + Send + 'static,
+    S: Service<I> + Send + Clone + 'static,
+    S::Error: Into<Error>,
+    S::Future: Send,
+{
+    type Response = S::Response;
+    type Error = Error;
+    type Future = Pin<Box<dyn Future<Output = Result<S::Response, Error>> + Send + 'static>>;
+
+    #[inline]
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx).map_err(Into::into)
+    }
+
+    fn call(&mut self, io: I) -> Self::Future {
+        let metrics = self.metrics.clone();
+        let inner = self.inner.clone();
+
+        Box::pin(async move {
+            metrics.inc_open();
+            match inner.oneshot(io).await.map_err(Into::into) {
```

Comment:
```suggestion

        metrics.inc_open();
        let call = self.inner.call(io);

        Box::pin(async move {
            match call.await.map_err(Into::into) {
```

The clone/oneshot invalidates the prior poll_ready -- poll_ready "reserves" us the ability to send a request. We want to use that reservation. When we clone, we invalidate that reservation. Oneshot combines waiting for a reservation and dispatching the request.

Reply:
@olix0r  Done in https://github.com/linkerd/linkerd2-proxy/pull/3355/commits/5ac135212f1a7f51d083d99bf7e7d7231d5d972d